File: HuHoBotPenguin-LLSE/lib/state.js
'use strict';

/**
 * 分群状态持久化：手动管理员 / 认证用户 / 管理方式 / 全量转发，逐群存档。
 * 对应 Java 版 command-state.ini（administrators / authenticated-users / administrator-modes / full-forwarding）。
 * 变更即写盘（tmp + rename 原子写）。
 */

const fs = require('fs');
const path = require('path');
const { root } = require('./config');

const log = typeof logger !== 'undefined' ? logger : console;

const MODE_QQ = 'QQ';
const MODE_MANUAL = 'MANUAL';
const MODE_BOTH = 'BOTH';

/** 配置值 qq/config/both → QQ/MANUAL/BOTH，对齐 Java AdminMode.from。 */
function mapConfigMode(value) {
    if (value === 'qq') return MODE_QQ;
    if (value === 'config' || value === 'manual') return MODE_MANUAL;
    return MODE_BOTH;
}

function isValidMode(value) {
    return value === MODE_QQ || value === MODE_MANUAL || value === MODE_BOTH;
}

class State {
    constructor(cfg) {
        this.cfg = cfg;
        this.file = path.join(root(), 'command-state.json');
        this.data = {
            administrators: {},
            'authenticated-users': {},
            'administrator-modes': {},
            'full-forwarding': {},
            bindings: {}
        };
        this.load();
    }

    load() {
        try {
            const raw = JSON.parse(fs.readFileSync(this.file, 'utf8'));
            if (raw.administrators && typeof raw.administrators === 'object') this.data.administrators = raw.administrators;
            if (raw['authenticated-users'] && typeof raw['authenticated-users'] === 'object') this.data['authenticated-users'] = raw['authenticated-users'];
            if (raw['administrator-modes'] && typeof raw['administrator-modes'] === 'object') this.data['administrator-modes'] = raw['administrator-modes'];
            if (raw['full-forwarding'] && typeof raw['full-forwarding'] === 'object') this.data['full-forwarding'] = raw['full-forwarding'];
            if (raw.bindings && typeof raw.bindings === 'object') this.data.bindings = raw.bindings;
        } catch (e) {
            // 首次运行没有状态文件，使用空状态
        }
    }

    save() {
        const tmp = this.file + '.tmp';
        try {
            fs.writeFileSync(tmp, JSON.stringify(this.data, null, 2) + '\n', 'utf8');
            fs.renameSync(tmp, this.file);
        } catch (e) {
            log.error('[HuHoBotPenguin] 状态写入失败：' + e.message);
        }
    }

    // ---- 手动管理员 ----

    _manualAdmins(group) {
        if (!this.data.administrators[group]) this.data.administrators[group] = [];
        return this.data.administrators[group];
    }

    listAdmins(group) {
        return this.data.administrators[group] || [];
    }

    containsManualAdmin(group, openid) {
        return this._manualAdmins(group).includes(openid);
    }

    addAdmin(group, openid) {
        const list = this._manualAdmins(group);
        if (!list.includes(openid)) {
            list.push(openid);
            this.save();
        }
    }

    removeAdmin(group, openid) {
        const list = this._manualAdmins(group);
        const index = list.indexOf(openid);
        if (index !== -1) {
            list.splice(index, 1);
            this.save();
        }
    }

    /** openid 是否在 手动管理员 + admin.openids 配置 的并集中。 */
    isManualAdmin(group, openid) {
        if (this.cfg.getList('admin.openids').includes(openid)) return true;
        return this.containsManualAdmin(group, openid);
    }

    // ---- 认证用户 ----

    _authenticated(group) {
        if (!this.data['authenticated-users'][group]) this.data['authenticated-users'][group] = [];
        return this.data['authenticated-users'][group];
    }

    isAuthenticated(group, openid) {
        return this._authenticated(group).includes(openid);
    }

    authenticate(group, openid) {
        const list = this._authenticated(group);
        if (!list.includes(openid)) {
            list.push(openid);
            this.save();
        }
    }

    revoke(group, openid) {
        const list = this._authenticated(group);
        const index = list.indexOf(openid);
        if (index !== -1) {
            list.splice(index, 1);
            this.save();
        }
    }

    // ---- 管理方式（每群覆盖，缺省取配置 admin.mode） ----

    mode(group) {
        const stored = this.data['administrator-modes'][group];
        if (stored && isValidMode(stored)) return stored;
        return mapConfigMode(this.cfg.getString('admin.mode', 'both'));
    }

    setMode(group, mode) {
        this.data['administrator-modes'][group] = mode;
        this.save();
    }

    /** 以角色 + 管理方式判定是否管理员（对齐 CommandSupport.requireAdmin）。 */
    isAdmin(group, openid, role) {
        const mode = this.mode(group);
        const roleOk = role === 'owner' || role === 'admin';
        const manualOk = this.isManualAdmin(group, openid);
        if (mode === MODE_QQ) return roleOk;
        if (mode === MODE_MANUAL) return manualOk;
        return roleOk || manualOk;
    }

    // ---- 全量转发（每群覆盖，缺省取配置 features.full-amount） ----

    isFullForwarding(group) {
        const value = this.data['full-forwarding'][group];
        if (value !== undefined) return !!value;
        return this.cfg.getBool('features.full-amount', false);
    }

    setFullForwarding(group, enabled) {
        this.data['full-forwarding'][group] = !!enabled;
        this.save();
    }

    // ---- 白名单绑定（QQ OpenID ⇄ 游戏名）：自助绑定 / 管理员解绑 ----

    _bindings(group) {
        if (!this.data.bindings) this.data.bindings = {};
        if (!this.data.bindings[group] || typeof this.data.bindings[group] !== 'object') {
            this.data.bindings[group] = {};
        }
        return this.data.bindings[group];
    }

    /** 查询 openid 在该群绑定的游戏名，无绑定返回 null。 */
    getBindingName(group, openid) {
        const groupMap = this.data.bindings && this.data.bindings[group];
        return groupMap && groupMap[openid] ? String(groupMap[openid]) : null;
    }

    bindName(group, openid, gameName) {
        this._bindings(group)[openid] = String(gameName);
        this.save();
        return String(gameName);
    }

    /** 解除 openid 在该群的绑定，返回原游戏名（无绑定返回 null）。 */
    unbindOpenid(group, openid) {
        const groupMap = this.data.bindings && this.data.bindings[group];
        if (!groupMap || !groupMap[openid]) return null;
        const gameName = String(groupMap[openid]);
        delete groupMap[openid];
        this.save();
        return gameName;
    }

    /** 按游戏名反查绑定它的 OpenID（管理员手动解绑用），未找到返回 null。 */
    findOpenidByGameName(group, gameName) {
        const groupMap = this.data.bindings && this.data.bindings[group];
        if (!groupMap) return null;
        const key = String(gameName).toLowerCase();
        for (const [openid, stored] of Object.entries(groupMap)) {
            if (String(stored).toLowerCase() === key) return openid;
        }
        return null;
    }
}

module.exports = { State, MODE_QQ, MODE_MANUAL, MODE_BOTH, mapConfigMode, isValidMode };

File: src/endstone_huhobot_penguin/state.py
"""
分群状态持久化：手动管理员 / 认证用户 / 管理方式 / 全量转发 / 白名单绑定，逐群存档。
对应 Java 版 command-state.ini。变更即写盘（tmp + rename 原子写）。
"""

import json
import os

from .logger import log

MODE_QQ = "QQ"
MODE_MANUAL = "MANUAL"
MODE_BOTH = "BOTH"


def map_config_mode(value):
    """配置值 qq/config/both → QQ/MANUAL/BOTH，对齐 Java AdminMode.from。"""
    if value == "qq":
        return MODE_QQ
    if value in ("config", "manual"):
        return MODE_MANUAL
    return MODE_BOTH


def is_valid_mode(value):
    return value in (MODE_QQ, MODE_MANUAL, MODE_BOTH)


class State:
    def __init__(self, root_dir, cfg):
        self.cfg = cfg
        self.file = os.path.join(root_dir, "command-state.json")
        self.data = {
            "administrators": {},
            "authenticated-users": {},
            "administrator-modes": {},
            "full-forwarding": {},
            "bindings": {},
        }
        self.load()

    def load(self):
        try:
            with open(self.file, "r", encoding="utf-8") as f:
                raw = json.load(f)
            if not isinstance(raw, dict):
                return
            for key in self.data:
                if isinstance(raw.get(key), dict):
                    self.data[key] = raw[key]
        except Exception:
            # 首次运行没有状态文件，使用空状态
            pass

    def save(self):
        tmp = self.file + ".tmp"
        try:
            with open(tmp, "w", encoding="utf-8") as f:
                json.dump(self.data, f, ensure_ascii=False, indent=2)
                f.write("\n")
            os.replace(tmp, self.file)
        except Exception as e:
            log.error("状态写入失败：" + str(e))

    # ---- 手动管理员 ----

    def _manual_admins(self, group):
        if group not in self.data["administrators"]:
            self.data["administrators"][group] = []
        return self.data["administrators"][group]

    def list_admins(self, group):
        return self.data["administrators"].get(group, [])

    def contains_manual_admin(self, group, openid):
        return openid in self._manual_admins(group)

    def add_admin(self, group, openid):
        lst = self._manual_admins(group)
        if openid not in lst:
            lst.append(openid)
            self.save()

    def remove_admin(self, group, openid):
        lst = self._manual_admins(group)
        if openid in lst:
            lst.remove(openid)
            self.save()

    def is_manual_admin(self, group, openid):
        """openid 是否在 手动管理员 + admin.openids 配置 的并集中。"""
        if openid in self.cfg.get_list("admin.openids"):
            return True
        return self.contains_manual_admin(group, openid)

    # ---- 认证用户 ----

    def _authenticated(self, group):
        if group not in self.data["authenticated-users"]:
            self.data["authenticated-users"][group] = []
        return self.data["authenticated-users"][group]

    def is_authenticated(self, group, openid):
        return openid in self._authenticated(group)

    def authenticate(self, group, openid):
        lst = self._authenticated(group)
        if openid not in lst:
            lst.append(openid)
            self.save()

    def revoke(self, group, openid):
        lst = self._authenticated(group)
        if openid in lst:
            lst.remove(openid)
            self.save()

    # ---- 管理方式（每群覆盖，缺省取配置 admin.mode） ----

    def mode(self, group):
        stored = self.data["administrator-modes"].get(group)
        if stored and is_valid_mode(stored):
            return stored
        return map_config_mode(self.cfg.get_string("admin.mode", "both"))

    def set_mode(self, group, mode):
        self.data["administrator-modes"][group] = mode
        self.save()

    def is_admin(self, group, openid, role):
        """以角色 + 管理方式判定是否管理员（对齐 CommandSupport.requireAdmin）。"""
        mode = self.mode(group)
        role_ok = role in ("owner", "admin")
        manual_ok = self.is_manual_admin(group, openid)
        if mode == MODE_QQ:
            return role_ok
        if mode == MODE_MANUAL:
            return manual_ok
        return role_ok or manual_ok

    # ---- 全量转发（每群覆盖，缺省取配置 features.full-amount） ----

    def is_full_forwarding(self, group):
        value = self.data["full-forwarding"].get(group)
        if value is not None:
            return bool(value)
        return self.cfg.get_bool("features.full-amount", False)

    def set_full_forwarding(self, group, enabled):
        self.data["full-forwarding"][group] = bool(enabled)
        self.save()

    # ---- 白名单绑定（QQ OpenID ⇄ 游戏名）：自助绑定 / 管理员解绑 ----

    def _bindings(self, group):
        if group not in self.data["bindings"] or not isinstance(self.data["bindings"][group], dict):
            self.data["bindings"][group] = {}
        return self.data["bindings"][group]

    def get_binding_name(self, group, openid):
        group_map = self.data["bindings"].get(group)
        if not group_map:
            return None
        value = group_map.get(openid)
        return str(value) if value else None

    def bind_name(self, group, openid, game_name):
        self._bindings(group)[openid] = str(game_name)
        self.save()
        return str(game_name)

    def unbind_openid(self, group, openid):
        group_map = self.data["bindings"].get(group)
        if not group_map or not group_map.get(openid):
            return None
        game_name = str(group_map[openid])
        del group_map[openid]
        self.save()
        return game_name

    def find_openid_by_game_name(self, group, game_name):
        group_map = self.data["bindings"].get(group)
        if not group_map:
            return None
        key = str(game_name).lower()
        for openid, stored in group_map.items():
            if str(stored).lower() == key:
                return openid
        return None
